docs/dev/truncate-with-tablets.md

```diff
@@ -0,0 +1,100 @@
+# Background
+
+The CQL TRUNCATE query is obviously data-modifying since it deletes user data, however,
+unlike DELETE queries, it is not based on a tombstone that carries a write-timestamp,
+with which we can determine for each cell, if it was logically written before
+or after the tombstone timestamp.
+
+Instead, truncate propagates in the cluster using an out-of-band RPC and is currently
+applied locally in each node using the local wall clock of each node (not synchronized
+with each other), typically after flushing the tables and taking a snapshot of them,
+if auto-snapshot is enabled.
+
+This presents a problem when tablets are enabled since there could be on-going tablet
+transitions in the background, that the user has no control over, like tablet migration,
+with which a tablet might escape truncate by migrating between nodes just at the wrong
+time so it won’t be considered for truncate by any node.
+
+This document lists the currently proposed solutions and lists the selected proposal.
+
+# Proposed solutions
+
+## 1. Short term mitigation
+
+We can disable any background load balancing operations (if any are running) and execute the
+truncate as it is currently implemented. We would issue a RPC which performs a memtable flush,
+then take a snapshot of the SSTables if auto-snapshot is enabled, and delete the SSTables
+belonging to the given table and associated views. After truncate is complete, we should
+re-enable load balancing.
+
+A problem with this solution is that a truncate operation which is issued while another
+truncate is already running could potentially run into the same problem with resurecting
+data which is in migration. Consider the following scenario:
+
+- a truncate table command is received (TT1)
+- load balancing is disabled as a result of TT1
+- truncating the table and views for TT1 is started
+- another truncate table command is received (TT2)
+- TT2 does not disable load balancing, as it is already disabled
+- truncating the table for TT1 completes 
+- TT1 enables load balancing
+- TT2 starts executing the truncate operation while load balancing is enabled potentially
+causing the problem with tablet migration resurecting truncated data
+
+Another potential problem is that load balancing can be manually enabled by the admin while
+truncate is still running, causing the same problem with resurrected data due to migration.
```

Comment:
A better short-term mitigation, which doesn't suffer from this issue, is what we did for repair: 98323be296fdc96c6f63e15fa196c6de0ae349af.

Reply:
I have added the description of the accepted implementation for safe truncate table with tablets.